## Exploring Hackers News Posts

In this project, I'll compare two different types of posts from Hacker News, a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts I'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

I'll specifically compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?
It should be noted that the data set I'm working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

In [1]:
import csv

f = open('hacker_news.csv')
hn = list(csv.reader(f))
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


## Removing Headers

In [2]:
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Now the header has been removed and the dataset is isolated and assigned to the variable hn.

## Extracting Ask HN and Show HN Posts

I have to identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [3]:
ask_posts = []
show_posts =[]
other_posts = []

for post in hn:
    title = post[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(post)
    elif title.lower().startswith("show hn"):
        show_posts.append(post)
    else:
        other_posts.append(post)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


I have identified 1744 "ask hn" posts and 1162 "show hn" posts.

## Calculate the Average Number of Comments for both the types of posts

In [4]:
total_ask_comments = 0

for post in ask_posts:
    total_ask_comments += int(post[4])
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [5]:
total_show_comments = 0

for post in show_posts:
    total_show_comments += int(post[4])
    
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


Ask posts defeat show posts 14-10 (approximately). Let's focus on ask posts.

## Are the amount of posts and comments influenced by the hour of the day?

In [6]:
import datetime as dt

result_list = []

for post in ask_posts:
    result_list.append(
        [post[6], int(post[4])]
    )

comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    comment = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

print(comments_by_hour)

{'14': 1416, '17': 1146, '11': 641, '21': 1745, '09': 251, '06': 397, '18': 1439, '20': 1722, '04': 337, '12': 687, '07': 267, '19': 1188, '02': 1381, '00': 447, '15': 4477, '03': 421, '05': 464, '10': 793, '22': 479, '23': 543, '13': 1253, '01': 683, '08': 492, '16': 1814}


In [7]:
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

print(avg_by_hour)

[['14', 13.233644859813085], ['17', 11.46], ['11', 11.051724137931034], ['21', 16.009174311926607], ['09', 5.5777777777777775], ['06', 9.022727272727273], ['18', 13.20183486238532], ['20', 21.525], ['04', 7.170212765957447], ['12', 9.41095890410959], ['07', 7.852941176470588], ['19', 10.8], ['02', 23.810344827586206], ['00', 8.127272727272727], ['15', 38.5948275862069], ['03', 7.796296296296297], ['05', 10.08695652173913], ['10', 13.440677966101696], ['22', 6.746478873239437], ['23', 7.985294117647059], ['13', 14.741176470588234], ['01', 11.383333333333333], ['08', 10.25], ['16', 16.796296296296298]]


In [9]:
def sort_list(a_list):
    a_list.sort(key = lambda x: x[1], reverse=True) 
    return a_list

sorted_avg_by_hour = sort_list(avg_by_hour)
print(sorted_avg_by_hour)

[['15', 38.5948275862069], ['02', 23.810344827586206], ['20', 21.525], ['16', 16.796296296296298], ['21', 16.009174311926607], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['18', 13.20183486238532], ['17', 11.46], ['01', 11.383333333333333], ['11', 11.051724137931034], ['19', 10.8], ['08', 10.25], ['05', 10.08695652173913], ['12', 9.41095890410959], ['06', 9.022727272727273], ['00', 8.127272727272727], ['23', 7.985294117647059], ['07', 7.852941176470588], ['03', 7.796296296296297], ['04', 7.170212765957447], ['22', 6.746478873239437], ['09', 5.5777777777777775]]


This way I have sorted the list of lists. The most favorable hour is the 15 (3 PM).

In [11]:
print("Top 5 Hours for 'Ask HN' Comments")
for hr, avg in sorted_avg_by_hour[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The most favorable time to post is the first afternoon or the evening.

## Calculate the Average Number of Points for both the types of posts

In [12]:
total_ask_points = 0

for post in ask_posts:
    total_ask_points += int(post[3])
    
avg_ask_points = total_ask_points / len(ask_posts)
print(avg_ask_points)

total_show_points = 0

for post in show_posts:
    total_show_points += int(post[3])
    
avg_show_points = total_show_points / len(show_posts)
print(avg_show_points)

15.061926605504587
27.555077452667813


Even in tems of points ask posts wins!
They defeat show posts 28-15 (approximately). Let's focus on ask posts.

## Is the amount of points influenced by the hour of the day?

In [17]:
import datetime as dt

result_points_list = []

for post in ask_posts:
    result_points_list.append(
        [post[6], int(post[3])]
    )

points_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_points_list:
    date = each_row[0]
    points = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in points_by_hour:
        points_by_hour[time] += points
    else:
        points_by_hour[time] = points

print(points_by_hour)

{'14': 1282, '17': 1941, '11': 825, '21': 1721, '09': 329, '06': 591, '18': 1741, '20': 1151, '04': 389, '12': 782, '07': 361, '19': 1513, '02': 793, '00': 451, '15': 3479, '03': 374, '05': 552, '10': 1102, '22': 511, '23': 581, '13': 2062, '01': 700, '08': 515, '16': 2522}


In [18]:
avg_points_by_hour = []

for hr in points_by_hour:
    avg_points_by_hour.append([hr, points_by_hour[hr] / counts_by_hour[hr]])

sorted_avg_points_by_hour = sort_list(avg_points_by_hour)
print(sorted_avg_points_by_hour)

[['15', 29.99137931034483], ['13', 24.258823529411764], ['16', 23.35185185185185], ['17', 19.41], ['10', 18.677966101694917], ['18', 15.972477064220184], ['21', 15.788990825688073], ['20', 14.3875], ['11', 14.224137931034482], ['19', 13.754545454545454], ['02', 13.672413793103448], ['06', 13.431818181818182], ['05', 12.0], ['14', 11.981308411214954], ['01', 11.666666666666666], ['08', 10.729166666666666], ['12', 10.712328767123287], ['07', 10.617647058823529], ['23', 8.544117647058824], ['04', 8.27659574468085], ['00', 8.2], ['09', 7.311111111111111], ['22', 7.197183098591549], ['03', 6.925925925925926]]


In [19]:
print("Top 5 Hours for 'Ask HN' Points")
for hr, avg in sorted_avg_points_by_hour[:5]:
    print(
        "{}: {:.2f} average points per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Points
15:00: 29.99 average points per post
13:00: 24.26 average points per post
16:00: 23.35 average points per post
17:00: 19.41 average points per post
10:00: 18.68 average points per post


Among the 'Ask HN' posts, the ones receiving more points are those posted in the first afternoon.

## Conclusions

If you want to receive points and comments as much as possible, you should post between 15:00 and 17:00 (3 PM - 5 PM), taking into account that I neglect posts without comments.